core/client/src/main/java/alluxio/client/block/stream/PacketInStream.java

```diff
@@ -0,0 +1,278 @@
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
+        PreconditionMessage.ERR_BUFFER_STATE.toString(), b.length, off, len);
+    if (len == 0) {
+      return 0;
+    }
+
+    readPacket();
+    if (mCurrentPacket == null) {
+      mEOF = true;
+    }
+    if (mEOF) {
+      close();
+      return -1;
+    }
+    int toRead = Math.min(len, mCurrentPacket.readableBytes());
+    mCurrentPacket.readBytes(b, off, toRead);
+    mPos += toRead;
+    return toRead;
+  }
+
+  @Override
+  public int positionedRead(long pos, byte[] b, int off, int len) throws IOException {
+    if (len == 0) {
+      return 0;
+    }
+    if (pos < 0 || pos >= mLength) {
+      return -1;
+    }
+
+    int lenCopy = len;
+    try (PacketReader reader = mPacketReaderFactory.create(pos, len)) {
+      // We try to read len bytes instead of returning after reading one packet because
+      // it is not free to create/close a PacketReader.
+      while (len > 0) {
+        ByteBuf buf = null;
+        try {
+          buf = reader.readPacket();
+          if (buf == null) {
+            break;
+          }
+          Preconditions.checkState(buf.readableBytes() <= len);
+          int toRead = buf.readableBytes();
+          buf.readBytes(b, off, toRead);
+          len -= toRead;
+          off += toRead;
+        } finally {
+          if (buf != null) {
+            buf.release();
+          }
+        }
+      }
+    }
+    if (lenCopy == len) {
+      return -1;
+    }
+    return lenCopy - len;
+  }
+
+  @Override
+  public long remaining() {
+    return mEOF ? 0 : mLength - mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(pos >= 0, PreconditionMessage.ERR_SEEK_NEGATIVE.toString(), pos);
+    Preconditions
+        .checkArgument(pos <= mLength, PreconditionMessage.ERR_SEEK_PAST_END_OF_REGION.toString(),
+            mId);
+    if (pos == mPos) {
+      return;
+    }
+    if (pos < mPos) {
+      mEOF = false;
+    }
+
+    closePacketReader();
+    mPos = pos;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    checkIfClosed();
+    if (n <= 0) {
+      return 0;
+    }
+
+    long toSkip = Math.min(remaining(), n);
+    mPos += toSkip;
+
+    closePacketReader();
+    return toSkip;
+  }
+
+  @Override
+  public void close() {
+    closePacketReader();
+    mClosed = true;
+  }
+
+  /**
+   * Reads a new packet from the channel if all of the current packet is read.
+   *
+   * @throws IOException if it fails to read the packet
+   */
+  private void readPacket() throws IOException {
+    if (mPacketReader == null) {
+      mPacketReader = mPacketReaderFactory.create(mPos, mLength - mPos);
+    }
+
+    if (mCurrentPacket != null && mCurrentPacket.readableBytes() == 0) {
+      destroyPacket(mCurrentPacket);
+      mCurrentPacket = null;
+    }
+    if (mCurrentPacket == null) {
+      mCurrentPacket = mPacketReader.readPacket();
```

Comment:
do we ever end up releasing ``mCurrentPacket` anywhere in this class? I could not find it.